sdk/core/azure-core/src/cryptography/md5.cpp

```diff
@@ -16,100 +16,66 @@
-            sizeof(hashLength),
-            &dataLength,
-            0);
-        if (!BCRYPT_SUCCESS(status))
-        {
-          throw std::runtime_error("BCryptGetProperty failed");
-        }
-        HashLength = hashLength;
-      }
-
-      ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
-    };
-
-    struct Md5HashContext
-    {
-      std::string buffer;
-      BCRYPT_HASH_HANDLE hashHandle = nullptr;
-      std::size_t hashLength = 0;
-    };
-  } // namespace _detail
+struct AlgorithmProviderInstance
+{
+  BCRYPT_ALG_HANDLE Handle;
+  std::size_t ContextSize;
+  std::size_t HashLength;
 
-  Md5Hash::Md5Hash()
+  AlgorithmProviderInstance()
   {
-    static _detail::AlgorithmProviderInstance AlgorithmProvider{};
-
-    _detail::Md5HashContext* md5Context = new _detail::Md5HashContext;
-    m_md5Context = md5Context;
-    md5Context->buffer.resize(AlgorithmProvider.ContextSize);
-    md5Context->hashLength = AlgorithmProvider.HashLength;
-
-    NTSTATUS status = BCryptCreateHash(
-        AlgorithmProvider.Handle,
-        &md5Context->hashHandle,
-        reinterpret_cast<PUCHAR>(&md5Context->buffer[0]),
-        static_cast<ULONG>(md5Context->buffer.size()),
-        nullptr,
-        0,
+    NTSTATUS status = BCryptOpenAlgorithmProvider(&Handle, BCRYPT_MD5_ALGORITHM, nullptr, 0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptOpenAlgorithmProvider failed");
+    }
+    DWORD objectLength = 0;
+    DWORD dataLength = 0;
+    status = BCryptGetProperty(
+        Handle,
+        BCRYPT_OBJECT_LENGTH,
+        reinterpret_cast<PBYTE>(&objectLength),
+        sizeof(objectLength),
+        &dataLength,
         0);
     if (!BCRYPT_SUCCESS(status))
     {
-      throw std::runtime_error("BCryptCreateHash failed");
+      throw std::runtime_error("BCryptGetProperty failed");
+    }
+    ContextSize = objectLength;
+    DWORD hashLength = 0;
+    status = BCryptGetProperty(
+        Handle,
+        BCRYPT_HASH_LENGTH,
+        reinterpret_cast<PBYTE>(&hashLength),
+        sizeof(hashLength),
+        &dataLength,
+        0);
+    if (!BCRYPT_SUCCESS(status))
+    {
+      throw std::runtime_error("BCryptGetProperty failed");
     }
+    HashLength = hashLength;
   }
 
-  Md5Hash::~Md5Hash()
-  {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
-    BCryptDestroyHash(md5Context->hashHandle);
-    delete md5Context;
-  }
+  ~AlgorithmProviderInstance() { BCryptCloseAlgorithmProvider(Handle, 0); }
+};
 
-  void Md5Hash::OnAppend(const uint8_t* data, std::size_t length)
-  {
-    _detail::Md5HashContext* md5Context = static_cast<_detail::Md5HashContext*>(m_md5Context);
+class Md5BCrypt : public Azure::Core::Cryptography::Hash {
+private:
+  std::string m_buffer;
+  BCRYPT_HASH_HANDLE m_hashHandle = nullptr;
+  std::size_t m_hashLength = 0;
+  AlgorithmProviderInstance m_algorithmProviderInstance;
```

Comment:
You must NOT make it a global static variable. `BCryptOpenAlgorithmProvider` cannot be called during dll initialization.

Reply:
And we should add a comment for that to remind future developers.

Reply:
Is this a global static?

Isn't it just a private field from an instance?

btw, we have the dll header to create global statics just fine  @antkmsft  added that.  But, I am confused here since I don't exepect it to be static global, but one per instance

Reply:
static variable defined in a function body is fine. Otherwise, it's global.

Reply:
> static variable defined in a function body is fine. Otherwise, it's global.

class A {
private:
  std::string m_someVar;
}

Are you telling me that `m_someVar` in previous example is global?

Reply:
It's not a `global` variable per se. But it has the same lifecycle as global variables.

Reply:
I am even more confused now.

The lifecycle of a private member is to init on instance cttor and die on instance dttor. Isn't it?
versus an static global variable which gets init at some point during app start up and gets destroyed at some point on app exit.

So if I do,

main() {

 {
   MyClass instance;
 }
}

The private members from `MyClass` init and die on its scope.

Reply:
Oh, okay, my mistake. I thought it was `static`.

Reply:
I'll mark this conversation as resolved.